Rename files written by PageSink

NikhilCollooru opened the pull request:
This PR enables the workers to send / receive metadata requests /responses from coordinator. 
If this feature is enabled for Hive, then all the files written by the HivePageSink for a given partition will have increasing whole numbers as file names. (For ex: Lets say this we enabled this feature and ds=2020-09-30 is written by HivePageSink. Then we can expect the file names of that partition to be 0,1,2,3..N etc ).

depended by https://github.com/facebookexternal/presto-facebook/pull/1125
```
== RELEASE NOTES ==
General Changes
* Add support for file renaming for Hive connector. This can be enabled with ``hive.file_renaming_enabled`` configuration property
```

highker commented:
rebase a bit maybe?

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorPageSinkProvider.java:
```diff
@@ -19,9 +19,11 @@
+
 public interface ConnectorPageSinkProvider
 {
-    ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorOutputTableHandle outputTableHandle, PageSinkProperties pageSinkProperties);
+    ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorOutputTableHandle outputTableHandle, Optional<ConnectorDistributedMetadataUpdater> distributedMetadataUpdater, PageSinkProperties pageSinkProperties);
 
-    ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorInsertTableHandle insertTableHandle, PageSinkProperties pageSinkProperties);
+    ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorInsertTableHandle insertTableHandle, Optional<ConnectorDistributedMetadataUpdater> distributedMetadataUpdater, PageSinkProperties pageSinkProperties);
```
I think we can directly inject `ConnectorDistributedMetadataUpdater` into the underlying page sink provider then passing through SPI? The design principle is to keep spi as simple as possible

NikhilCollooru replied:
Got it.

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java:
```diff
@@ -700,6 +702,14 @@ default void revokeTablePrivileges(ConnectorSession session, SchemaTableName tab
+    /**
+     * Handles metadata update requests and sends the results back to worker
+     */
+    default List<MetadataUpdateResult> updateMetadata(List<MetadataUpdateRequest> metadataUpdateRequests, String queryId)
```
```java
ConnectorMetadataUpdateHandle updateMetadata(ConnectorSession session, ConnectorMetadataUpdateHandle handle)
```

Likely you can squeeze input and output into one object. But if not, feel free to create two handle interfaces

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/MetadataUpdateRequest.java:
```diff
@@ -0,0 +1,21 @@
+ */
+package com.facebook.presto.spi;
+
+public interface MetadataUpdateRequest
```
`ConnectorMetadataUpdateHandle`. We don't need to define anything in the field. Check how other handle interfaces are defined.

NikhilCollooru replied:
We need to create update request object before sending to coordinator. To create that object we need `ConnectorMetadataUpdateHandle` to have basic interface functions right ?

highker replied:
All these functions are called hive connector. We usually don't put anything in engine level. If we wanna delegate to underlying connectors, we do downcast within its connector.

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorDistributedMetadataUpdater.java:
```diff
@@ -0,0 +1,33 @@
+
+public interface ConnectorDistributedMetadataUpdater
+{
+    long addMetadataUpdateRequest(int count);
+
+    List<MetadataUpdateRequest> getPendingMetadataUpdateRequests();
+
+    void removeResultFuture(long requestId);
+
+    CompletableFuture<List<String>> getMetadataResult(long requestID);
+
+    void setMetadataUpdateResults(List<MetadataUpdateResult> results);
```
We only need to expose interfaces that are accessed at engine level. For all that are used by connectors only, we don't need to expose them.

highker commented on presto-spi/src/main/java/com/facebook/presto/spi/MetadataUpdateResult.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.List;
+
+public interface MetadataUpdateResult
+{
+    long getRequestId();
+
+    List<String> getUpdates();
+}
```
same

NikhilCollooru commented:
> High-level comments of this PR: we need tests; a lot of unit tests and integration tests. Probably enable the feature for all write tests for hive connector. We also need to make sure the file names are in sequential order.

Enabling this feature will make the writer wait for fileName to be sent by coordinator. I don't think the normal write tests for hive connector have the setup to do communication with coordinator. I will check if I can add integration test.

NikhilCollooru commented on presto-main/src/main/java/com/facebook/presto/execution/TaskMetadataContext.java:
```diff
@@ -0,0 +1,57 @@
+        this.metadataUpdaters = new CopyOnWriteArrayList<>();
+    }
+
+    public void setConnectorId(ConnectorId id)
```
I made `TaskMetadataContext` an optional parameter in `TaskContext` and initiated it inside `TableWriterOperatorFactory`, as we can know the `connectorId` only when we reach there.

NikhilCollooru commented:
> High-level comments of this PR: we need tests; a lot of unit tests and integration tests. Probably enable the feature for all write tests for hive connector. We also need to make sure the file names are in sequential order.

I agree. I added `TestHiveFileRenamer` to test the sequential order of the fileNames. I also added `TestHiveMetadataUpdater` and `TestHiveMetadataUpdateHandle`. Will try and check if more tests can be added.

highker commented on presto-benchmark/src/main/java/com/facebook/presto/benchmark/AbstractOperatorBenchmark.java:
```diff
@@ -224,7 +224,8 @@ protected final OperatorFactory createHashProjectOperator(int operatorId, PlanNo
 
         PageFunctionCompiler functionCompiler = new PageFunctionCompiler(localQueryRunner.getMetadata(), 0);
-        projections.add(new PageProjectionWithOutputs(functionCompiler.compileProjection(session.getSqlFunctionProperties(), translatedHashExpression, Optional.empty()).get(), new int[] {types.size()}));
+        projections.add(new PageProjectionWithOutputs(functionCompiler.compileProjection(session.getSqlFunctionProperties(), translatedHashExpression, Optional.empty()).get(), new int[] {
```
one param per line for function `compileProjection`

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java:
```diff
@@ -124,6 +125,8 @@ public void configure(Binder binder)
         binder.bind(HiveWriterStats.class).in(Scopes.SINGLETON);
         newExporter(binder).export(HiveWriterStats.class).as(generatedNameOf(HiveWriterStats.class, connectorId));
 
+        binder.bind(HiveMetadataUpdaterProvider.class).in(Scopes.SINGLETON);
```
I guess you don't need this as it has already been bound below.

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataUpdateHandle.java:
```diff
@@ -0,0 +1,106 @@
+        implements ConnectorMetadataUpdateHandle
+{
+    private final String requestId;
+    private final String schemaName;
+    private final String tableName;
```
Use `SchemaTableName`

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataUpdateHandle.java:
```diff
@@ -0,0 +1,106 @@
+public class HiveMetadataUpdateHandle
+        implements ConnectorMetadataUpdateHandle
+{
+    private final String requestId;
```
Just use `UUID` maybe

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataUpdateHandle.java:
```diff
@@ -0,0 +1,106 @@
+    private final String requestId;
+    private final String schemaName;
+    private final String tableName;
+    private final String partitionName;
```
Use `Optional<String>`; same for the one below

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataUpdateHandle.java:
```diff
@@ -0,0 +1,106 @@
+    private final String schemaName;
+    private final String tableName;
+    private final String partitionName;
+    private final String metadataUpdate;
```
`fileName`. Also add a comment to explain when it would be non-null

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataUpdateHandle.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @JsonProperty
+    public String getMetadataUpdate()
+    {
+        return metadataUpdate;
+    }
```
```java
    public Optional<String> getMetadataUpdate()
    {
        return fileName;
    }

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataUpdateHandle.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        HiveMetadataUpdateHandle o = (HiveMetadataUpdateHandle) obj;
+        return requestId.equals(o.requestId) &&
+                schemaName.equals(o.schemaName) &&
+                tableName.equals(o.tableName) &&
+                Objects.equals(partitionName, o.partitionName);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(requestId, schemaName, tableName, partitionName);
+    }
```
Seems missing a field here.

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataUpdater.java:
```diff
@@ -0,0 +1,89 @@
+            HiveMetadataUpdateHandle updateResult = (HiveMetadataUpdateHandle) connectorMetadataUpdateHandle;
+            String requestId = updateResult.getRequestId();
+
+            if (!requestFutureMap.containsKey(requestId)) {
+                continue;
+            }
+
+            requestFutureMap.get(requestId).set(updateResult.getMetadataUpdate());
```
Is this thread-safe? Wonder what would happen if two threads set the same future. Might be fine, but worth digging and comment here.

highker commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHiveMetadataUpdater.java:
```diff
@@ -0,0 +1,76 @@
+import static com.facebook.presto.hive.TestHiveMetadataUpdateHandle.TEST_TABLE;
+import static org.testng.Assert.assertEquals;
+
+public class TestHiveMetadataUpdater
```
Shall we have more multi-threading tests for this class?

NikhilCollooru replied:
As per our new design , we have one `HiveMetadataUpdater` instance per each task. On other hand we need multi-threaded tests for `HiveFileRenamer` as we have a singleton on coordinator for all tasks. So added multi-threaded tests for `HiveFileRenamer`

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataUpdater.java:
```diff
@@ -0,0 +1,89 @@
+public class HiveMetadataUpdater
+        implements ConnectorMetadataUpdater
+{
+    private final Map<String, SettableFuture<String>> requestFutureMap = new ConcurrentHashMap<>();
+    private final Map<Integer, String> writerRequestMap = new ConcurrentHashMap<>();
+    private final Queue<HiveMetadataUpdateHandle> hiveMetadataRequestQueue = new ConcurrentLinkedQueue<>();
```
Let's put some comments to the data structures. BTW, seems we don't need a `Queue` type for `hiveMetadataRequestQueue` because there is no order. But let's keep that because we might come back to `getPendingMetadataUpdateRequests` and optimize it. The current impl for `getPendingMetadataUpdateRequests` is a bit brute force.

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataUpdater.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    @Override
+    public void setMetadataUpdateResults(List<ConnectorMetadataUpdateHandle> results)
```
This looks like an expensive call directly executed from HTTP thread (from `TaskResource.updateMetadataResults`). Potentially, this will exhaust jetty threads. A good way is to make it async and move it to a CPU thread.

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSinkProvider.java:
```diff
@@ -113,17 +114,21 @@ public HivePageSinkProvider(
         HiveWritableTableHandle handle = (HiveOutputTableHandle) tableHandle;
-        return createPageSink(handle, true, session, pageSinkProperties.isCommitRequired());
+        HiveMetadataUpdater hiveMetadataUpdater = (HiveMetadataUpdater) pageSinkProperties.getMetadataUpdater().orElse(null);
+        checkArgument(hiveMetadataUpdater != null, "Metadata Updater for HivePageSink is null");
```
use `isPresent` rather than making it null

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSinkProvider.java:
```diff
@@ -113,17 +114,21 @@ public HivePageSinkProvider(
     public ConnectorPageSink createPageSink(ConnectorTransactionHandle transaction, ConnectorSession session, ConnectorOutputTableHandle tableHandle, PageSinkContext pageSinkProperties)
```
`PageSinkContext pageSinkProperties`: seems you forgot to change the param names from the type renaming

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataUpdaterProvider.java:
```diff
@@ -0,0 +1,32 @@
+public class HiveMetadataUpdaterProvider
+        implements ConnectorMetadataUpdaterProvider
+{
+    @Inject
+    public HiveMetadataUpdaterProvider() {}
```
Remove this. Put `@Inject` at class definition

highker reviewed (commented):
still under review. Let's separate the first 3 commits into another PR. Looks like they are good mostly as a standalone piece.

NikhilCollooru commented:
Created a separate PR https://github.com/prestodb/presto/pull/15122  for the first 3 commits as suggested

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java:
```diff
@@ -339,6 +347,14 @@ private void writePage(Page page)
 
+    private void sendMetadataUpdateRequest(Optional<String> partitionName, int writerIndex)
+    {
+        if (bucketFunction != null) {
```
add a comment to this `if`

highker commented on presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java:
```diff
@@ -313,6 +318,24 @@ private TaskStats getTaskStats(TaskHolder taskHolder)
         return new TaskStats(taskStateMachine.getCreatedTime(), endTime);
     }
 
+    public MetadataUpdates getMetadataUpdateRequests(TaskHolder taskHolder)
```
private

highker commented on presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java:
```diff
@@ -313,6 +318,24 @@ private TaskStats getTaskStats(TaskHolder taskHolder)
+    public MetadataUpdates getMetadataUpdateRequests(TaskHolder taskHolder)
+    {
+        ConnectorId connectorId = null;
+        List<ConnectorMetadataUpdateHandle> connectorMetadataUpdates = new ArrayList<>();
```
Use list builder

highker commented on presto-main/src/main/java/com/facebook/presto/metadata/MetadataUpdates.java:
```diff
@@ -0,0 +1,60 @@
+{
+    public static final MetadataUpdates DEFAULT_METADATA_UPDATES = new MetadataUpdates(ImmutableList.of());
+
+    private ConnectorId connectorId;
+    private List<ConnectorMetadataUpdateHandle> metadataUpdates;
```
both are final

highker reviewed (commented):
"Piggyback MetadataUpdateRequests to coordinator as part of TaskInfo" LGTM % nits

highker reviewed (commented):
"Add new POST {taskId}/metadataresults endpoint on worker" LGTM

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java:
```diff
@@ -228,6 +238,19 @@ public long getValidationCpuNanos()
+                ListenableFuture<?> fileNameFuture = toListenableFuture(hiveMetadataUpdater.getMetadataResult(i));
+                SettableFuture renamingFuture = SettableFuture.create();
+                futures.add(renamingFuture);
+                int writerIndex = i;
```
this assignment looks unnecessary

NikhilCollooru replied:
If i use `i` as input param to lambda function then i am getting syntax error `Variable used in lambda expression should be final or effectivelyfinal`. Hence added this line.

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java:
```diff
@@ -228,6 +238,19 @@ public long getValidationCpuNanos()
+        if (waitForFileRenaming && verificationTasks.isEmpty()) {
+            ImmutableList.Builder<Slice> partitionUpdatesWithRenamedFileNames = ImmutableList.builder();
+            List<ListenableFuture<?>> futures = new ArrayList<>();
+            for (int i = 0; i < writers.size(); i++) {
```
s/i/writerIndex

NikhilCollooru replied:
Like i said above. If i use `i` as input param to lambda function then i am getting syntax error `Variable used in lambda expression should be final or effectivelyfinal`. Hence used `i` here and assigned it a new int variable `writerIndex` below.

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java:
```diff
@@ -353,6 +376,41 @@ private void sendMetadataUpdateRequest(Optional<String> partitionName, int write
+        PartitionUpdate partitionUpdate = writer.getPartitionUpdate();
+
+        // Only one file is written by a writer
+        FileWriteInfo fileWriteInfo = partitionUpdate.getFileWriteInfos().get(0);
```
assert list size is one

highker reviewed (commented):
"Rename files written by HivePageSink": make rename async

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -2879,6 +2883,12 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
     }
 
+    @Override
+    public List<ConnectorMetadataUpdateHandle> getMetadataUpdateResults(List<ConnectorMetadataUpdateHandle> metadataUpdateRequests, String queryId)
```
Use `QueryId` as the type for the second param

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveFileRenamer.java:
```diff
@@ -0,0 +1,123 @@
+
+public class HiveFileRenamer
+{
+    private final LoadingCache<HiveMetadataUpdateKey, AtomicLong> partitionFileCounterCache;
+    private final LoadingCache<HiveMetadataUpdateHandle, String> hiveMetadataResultCache;
```
We should use a concurrent hash map or something like to to remember everything. We purged finished queries by adding a callback to `QueryManager.addStateChangeListener`. Note that the callback itself could unfortunately introduce another interface in connector metadata. Let me know if you might come up with better solution than that.

NikhilCollooru replied:
How about we inject `QueryManager` directly into `HiveFileRenamer` because this is to help purges entries in `HiveFileRenamer` specifically. WDYT ?

highker reviewed (commented):
"POST the metadata results to worker": we need good ways to purge queries on coordinators

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:
```diff
@@ -181,6 +181,10 @@
     private boolean isPartialAggregationPushdownForVariableLengthDatatypesEnabled;
 
+    private boolean fileRenamingEnabled;
+    private long fileRenamerCacheSize = 1000000;
+    private Duration fileRenamerCacheTtl = new Duration(15, TimeUnit.MINUTES);
```
We don't need this if we have good way to purge queries

highker reviewed (commented):
"Add config parameters to enable hive file renaming": LGTM

highker reviewed (commented):
"Disable listFiles call to underlying storage": let's remove this commit in this PR. That is more like the standalone piece to review in the next PR

highker commented:
First pass done. Feel free to re-request my reviews once the comments are addressed.

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveFileRenamer.java:
```diff
@@ -0,0 +1,115 @@
+
+public class HiveFileRenamer
+{
+    private final ConcurrentMap<QueryId, Map<HiveMetadataUpdateKey, AtomicLong>> queryPartitionFileCounterMap = new ConcurrentHashMap<>();
+    private final ConcurrentMap<QueryId, Map<HiveMetadataUpdateHandle, String>> queryHiveMetadataResultMap = new ConcurrentHashMap<>();
```
`Map<...>`

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveFileRenamer.java:
```diff
@@ -0,0 +1,115 @@
+
+    public List<ConnectorMetadataUpdateHandle> getMetadataUpdateResults(List<ConnectorMetadataUpdateHandle> metadataUpdateRequests, QueryId queryId)
+    {
+        List<ConnectorMetadataUpdateHandle> metadataUpdateResults = new ArrayList<>();
```
use immutable list builder

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveFileRenamer.java:
```diff
@@ -0,0 +1,115 @@
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class HiveFileRenamer
```
Let's add a test to `TestHiveDistributedQueries` or any other integration tests to make sure after test ends, there is no dangling queries in the map. #14784 is an example.

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveFileRenamer.java:
```diff
@@ -0,0 +1,115 @@
+        return fileName;
+    }
+
+    private class HiveMetadataUpdateKey
```
static

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveFileRenamer.java:
```diff
@@ -0,0 +1,115 @@
+        }
+
+        HiveMetadataUpdateKey key = new HiveMetadataUpdateKey(request);
+        partitionFileCounterMap.putIfAbsent(key, new AtomicLong());
```
Do we wanna start with `new AtomicLong(1)`? Actually, check the bucketed case. If bucketed file name started with `1` then we can use `AtomicLong(1)`; otherwise `AtomicLong(0)`.

NikhilCollooru replied:
In bucketed case the file name starts with 0. `AtomicLong()` is same as `AtomicLong(0)` . But will change it to `AtomicLong(0)` as suggested, to be on safe side.

highker commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileRenamer.java:
```diff
@@ -0,0 +1,226 @@
+@Test
+public class TestHiveFileRenamer
+{
+    private static final QueryId TEST_QUERYID = new QueryId("test");
```
`TEST_QUERY_ID`

highker commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileRenamer.java:
```diff
@@ -0,0 +1,226 @@
+        List<String> aggregatedFileNames = new ArrayList<>();
+        List<ConnectorMetadataUpdateHandle> requests = createHiveMetadataUpdateRequests(TEST_SCHEMA_NAME, TEST_TABLE_NAME, TEST_PARTITION_NAME);
+        List<String> fileNames = getFileNames(hiveFileRenamer, requests);
+        aggregatedFileNames.addAll(fileNames);
```
`List<String> aggregatedFileNames = new ArrayList<>(fileNames);`

highker commented on presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java:
```diff
@@ -138,6 +141,7 @@
     private final TransactionManager transactionManager;
 
     private final ConcurrentMap<String, Collection<ConnectorMetadata>> catalogsByQueryId = new ConcurrentHashMap<>();
+    private final Set<QueryId> queriesWithRegisteredCallbacks = new HashSet<>();
```
Do we actually need this?

NikhilCollooru replied:
This is to make sure that we add only 1 callback for a query.

highker commented on presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java:
```diff
@@ -1210,6 +1214,28 @@ public void revokeTablePrivileges(Session session, QualifiedObjectName tableName
+        ConnectorId connectorId = metadataUpdateRequests.getConnectorId();
+        ConnectorMetadata metadata = getCatalogMetadata(session, connectorId).getMetadata();
+
+        if (queryManager != null && !queriesWithRegisteredCallbacks.contains(queryId)) {
```
It's ok to remove multiple times as long as we can make sure they are cleaned up

NikhilCollooru replied:
We want to add only 1 callback for a given queryId. If we don't check, we will end up adding callback everytime we call getMetadataUpdateResults. That might result is adding 10's of callbacks just for 1 query.

highker commented on presto-main/src/main/java/com/facebook/presto/server/remotetask/TaskInfoFetcher.java:
```diff
@@ -225,6 +250,11 @@ private synchronized void sendNextRequest()
         }
 
+        MetadataUpdates metadataUpdateRequests = taskInfo.getMetadataUpdates();
+        if (metadataUpdateRequests.getMetadataUpdates().size() > 0) {
```
`if (!metadataUpdateRequests.getMetadataUpdates().isEmpty())`

highker reviewed (commented):
Overall look good to me. Minor comments only

highker commented on presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java:
```diff
@@ -313,6 +317,24 @@ private TaskStats getTaskStats(TaskHolder taskHolder)
+
+        if (taskHolder.getTaskExecution() != null) {
+            TaskMetadataContext taskMetadataContext = taskHolder.getTaskExecution().getTaskContext().getTaskMetadataContext();
+            if (taskMetadataContext.getMetadataUpdaters().size() > 0) {
```
`!taskMetadataContext.getMetadataUpdaters().isEmpty()`

highker commented on presto-main/src/main/java/com/facebook/presto/execution/TaskManager.java:
```diff
@@ -142,4 +143,11 @@ TaskInfo updateTask(
+    /**
+     * Update the results of metadata requests sent
+     */
+    default void updateMetadataResults(TaskId taskId, MetadataUpdates metadataUpdates)
```
remove default; throw exception for all other implementations.

highker commented on presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java:
```diff
@@ -138,6 +141,7 @@
     private final TransactionManager transactionManager;
 
     private final ConcurrentMap<String, Collection<ConnectorMetadata>> catalogsByQueryId = new ConcurrentHashMap<>();
+    private final Set<QueryId> queriesWithRegisteredCallbacks = new HashSet<>();
```
`ConcurrentHashMap.newKeySet();`

highker reviewed (approved):
good work! Nits only
